test/serialize.jl

```diff
@@ -280,16 +280,34 @@ create_serialization_stream() do s
 end
 
 # cycles
+module CycleFoo
+    echo(x)=x
+end
 create_serialization_stream() do s
-    A = Any[1,2,3,4,5]
+    echo(x) = x
+    afunc = (x)->x
+    A = Any[1,2,3,abs,abs,afunc,afunc,echo,echo,CycleFoo.echo,CycleFoo.echo,4,5]
     A[3] = A
     serialize(s, A)
     seekstart(s)
     b = deserialize(s)
     @test b[3] === b
     @test b[1] == 1
-    @test b[5] == 5
-    @test length(b) == 5
+    @test b[4] === abs
+    @test b[5] === b[4]
+    @test b[5](-1) == 1
+
+    @test b[6] === b[7]
+    @test b[6]("Hello") == "Hello"
+
+    @test b[8] === b[9]
+    @test b[8]("World") == "World"
+
+    @test b[10] === b[11]
+    @test b[10]("foobar") == "foobar"
+
+    @test b[end] == 5
+    @test length(b) == length(A)
```

Comment:
As the patch stands, deserialized `A` is 

```
1 ====> 1
2 ====> 2
3 ====> Any[1,2,#= circular reference =#,abs,abs,(anonymous function),AST(:($(Expr(:lambda, Any[:(x::Any)], Any[Any[Any[:x,:Any,0]],Any[],0,Any[]], :(begin  # none, line 1:
        return x
    end))))),echo,:(begin  # none, line 1:
        return x
    end),CycleFoo.echo,:(return x),4,5]
4 ====> abs
5 ====> abs
6 ====> (anonymous function)
7 ====> AST(:($(Expr(:lambda, Any[:(x::Any)], Any[Any[Any[:x,:Any,0]],Any[],0,Any[]], :(begin  # none, line 1:
        return x
    end)))))
8 ====> echo
9 ====> begin  # none, line 1:
    return x
end
10 ====> CycleFoo.echo
11 ====> return x
12 ====> 4
13 ====> 5
```

indexexes 6,7 should be the same. As also index pairs (8,9) and (10,11)